include/swift/Syntax/Serialization/SyntaxDeserialization.h

```diff
@@ -0,0 +1,211 @@
+  static void enumeration(IO &in, swift::tok &value) {
+#define TOKEN(name) in.enumCase(value, #name, swift::tok::name);
+#include "swift/Syntax/TokenKinds.def"
+  }
+};
+
+/// Deserialization traits for Trivia.
+/// An array of the underlying TriviaPieces will be deserialized as Trivia.
+template <typename Context>
+swift::syntax::TriviaPiece yamlize(IO &io, Context &Ctx) {
+  io.beginMapping();
+  auto ret = MappingTraits<swift::syntax::TriviaPiece>::mapping(io);
+  io.endMapping();
+  return ret;
+}
+
+template <typename Context>
+void yamlize(IO &io, std::vector<swift::syntax::TriviaPiece> &Seq, bool,
+             Context &Ctx) {
+  unsigned incnt = io.beginSequence();
+  for (unsigned i = 0; i < incnt; ++i) {
+    void *SaveInfo;
+    if (io.preflightElement(i, SaveInfo)) {
+      Seq.push_back(yamlize(io, Ctx));
+      io.postflightElement(SaveInfo);
+    }
+  }
+  io.endSequence();
+}
+
+template <> struct SequenceTraits<std::vector<swift::syntax::TriviaPiece>> {
+  static size_t size(IO &in, std::vector<swift::syntax::TriviaPiece> &seq) {
+    return seq.size();
+  }
+};
+
+/// Deserialization traits for RawSyntax list.
+template <> struct SequenceTraits<std::vector<swift::RC<swift::RawSyntax>>> {
+  static size_t size(IO &in, std::vector<swift::RC<swift::RawSyntax>> &seq) {
+    return seq.size();
+  }
+  static swift::RC<swift::RawSyntax> &
+  element(IO &in, std::vector<swift::RC<swift::RawSyntax>> &seq, size_t index) {
+    if (seq.size() <= index) {
+      seq.resize(index + 1);
+    }
+    return const_cast<swift::RC<swift::RawSyntax> &>(seq[index]);
+  }
+};
+
+/// An adapter struct that provides a nested structure for token content.
+struct TokenDescription {
+  bool hasValue = false;
+  swift::tok Kind;
+  StringRef Text;
+};
+
+/// Desrialization traits for TokenDescription.
+/// TokenDescriptions always serialized with a token kind, which is
+/// the stringified version of their name in the tok:: enum.
+/// ```
+/// {
+///   "kind": <token name, e.g. "kw_struct">,
+/// }
+/// ```
+///
+/// For tokens that have some kind of text attached, like literals or
+/// identifiers, the serialized form will also have a "text" key containing
+/// that text as the value.
+template <> struct MappingTraits<TokenDescription> {
+  static void mapping(IO &in, TokenDescription &value) {
+    in.mapRequired("kind", value.Kind);
+    value.hasValue = true;
+    switch (value.Kind) {
+    case swift::tok::contextual_keyword:
+    case swift::tok::integer_literal:
+    case swift::tok::floating_literal:
+    case swift::tok::string_literal:
+    case swift::tok::unknown:
+    case swift::tok::code_complete:
+    case swift::tok::identifier:
+    case swift::tok::oper_binary_unspaced:
+    case swift::tok::oper_binary_spaced:
+    case swift::tok::oper_postfix:
+    case swift::tok::oper_prefix:
+    case swift::tok::dollarident:
+    case swift::tok::comment:
+    case swift::tok::string_segment:
+      in.mapRequired("text", value.Text);
```

Comment:
I'm split here on whether we should use `mapOptional("text", value.Text)` here instead of switching over token types. @nkcsgexi, @rintaro thoughts?

Reply:
Yeah, I think manually typing these token kinds can be error-prone. Although we're not so likely to add new token kinds in the future, I prefer we avoid duplicating token's meta data. @DexinLi , could you try to generate this function, or any sub-function of it, say `shouldTokenRequireText()` in `SyntaxKind.h.gyb`?

Reply:
@nkcsgexi  `shouldTokenRequireText()` should be defined in [TokenKinds.def](https://github.com/apple/swift/blob/master/include/swift/Syntax/TokenKinds.def)  :P
@harlanhaskins  This part was copied from the [serializer](https://github.com/apple/swift/blob/master/include/swift/Syntax/Serialization/SyntaxSerialization.h#L92).

Since `shouldTokenRequireText()` is only used in syntax de/serializer, I think `mapOptional`  would be a good choice. Only need to maintain the switching part in serializer.